### PR TITLE
Implement `gpu_kernel_multiple_outputs`

### PR DESCRIPTION
**Pull request by crcrpar**

### CODE REVIEW

File: aten/src/ATen/native/cuda/MemoryAccess.cuh [hunk 1]
```diff
@@ -193,6 +222,54 @@ struct vectorized {
   }
 };
 
+template <int num_outputs, typename data_t, typename dtypes_t, typename offset_calc_t>
+struct multi_outputs_unroll {
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/MemoryAccess.cuh:226]: Instead of copy-pasting the implementation, can you modify the original in place to allow multiple outputs? Or another approach could be to make this a subclass of `unroll`.

File: aten/src/ATen/native/cuda/CUDALoops.cuh [hunk 2]
```diff
@@ -296,4 +313,34 @@ void gpu_kernel_impl(TensorIterator& iter, const func_t& f) {
...
+  TORCH_INTERNAL_ASSERT(iter.ntensors() == ntensors);
+
+  at::detail::Array<char*, ntensors> data;
+  for (int i = 0; i < ntensors; i++) {
+    data[i] = (char*)iter.data_ptr(i);
+  }
+
+  int64_t numel = iter.numel();
+
+  at::detail::Array<ScalarType, ntensors> dtypes;
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/CUDALoops.cuh:336]: Is this needed?

File: aten/src/ATen/native/cuda/Loops.cuh [hunk 3]
```diff
@@ -37,16 +37,16 @@ struct needs_dynamic_casting<func_t, 0> {
   }
 };
 
-template<int N>
+template<int N, int num_outputs = 1>
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/Loops.cuh:40]: number of output can be obtained from `iter`

File: aten/src/ATen/native/cuda/CUDALoops.cuh [hunk 4]
```diff
@@ -296,4 +313,34 @@ void gpu_kernel_impl(TensorIterator& iter, const func_t& f) {
...
+  for (int i = 0; i < ntensors; i++) {
+    dtypes[i] = iter.tensor(i).scalar_type();
+  }
+
+  auto input_calc = make_input_offset_calculator<num_inputs, num_outputs>(iter);
+  auto output_calc = ::make_offset_calculator<num_outputs>(iter);
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/CUDALoops.cuh:342]: I think you will need to use `make_output_offset_calculator` otherwise you will be using bytes as offset.

File: aten/src/ATen/native/cuda/CUDALoops.cuh [hunk 5]
```diff
@@ -151,7 +151,7 @@ __device__ inline void elementwise_kernel_helper(func_t f, policy_t policy) {
   #pragma unroll
   for (int i = 0; i < thread_work_size; i++) {
     if (policy.check_inbounds(i)) {
-      results[i] = c10::guts::apply(f, args[i]);
+      policy.store_result(i, c10::guts::apply(f, args[i]), results);
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/CUDALoops.cuh:154]: Why are you doing memory access here? There is a `policy.store(results, idx);` in the end.

File: aten/src/ATen/native/cuda/MemoryAccess.cuh [hunk 6]
```diff
@@ -191,6 +224,38 @@ struct vectorized {
...
+};
+
+template <typename data_t, typename inp_calc_t, typename out_calc_t, int num_outputs>
+struct multi_outputs_unroll : unroll<data_t, inp_calc_t, out_calc_t, num_outputs> {
+
+  __device__ multi_outputs_unroll(data_t data, int remaining, inp_calc_t ic, out_calc_t oc):
+    unroll<data_t, inp_calc_t, out_calc_t, num_outputs>({data, remaining, ic, oc}) {}
+
+  template <typename return_t>
+  __device__ inline void store(return_t *from, int idx) {
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/MemoryAccess.cuh:241]: what is `store` and what is `store_result`?

File: aten/src/ATen/native/cuda/CUDALoops.cuh (hunk 5 above)
Comment (zasdfgbnm) [reply aten/src/ATen/native/cuda/CUDALoops.cuh:154]: Here, would `return_t` be a tuple, and `results[i] = c10::guts::apply(f, args[i]);` would just correctly store the tuple on registers? Why it needs change?

File: aten/src/ATen/native/cuda/CUDALoops.cuh (hunk 2 above)
Comment (crcrpar) [reply aten/src/ATen/native/cuda/CUDALoops.cuh:336]: No, this isn't. thank you.

File: aten/src/ATen/native/cuda/Loops.cuh (hunk 3 above)
Comment (crcrpar) [reply aten/src/ATen/native/cuda/Loops.cuh:40]: thank you.

File: aten/src/ATen/native/cuda/CUDALoops.cuh (hunk 5 above)
Comment (crcrpar) [reply aten/src/ATen/native/cuda/CUDALoops.cuh:154]: Just because it doesn't work if `return_t` is `tuple`.

Ref: https://godbolt.org/z/uvcy8G

File: aten/src/ATen/native/cuda/MemoryAccess.cuh (hunk 6 above)
Comment (crcrpar) [reply aten/src/ATen/native/cuda/MemoryAccess.cuh:241]: `store` does store the results from `* from` to `data`. `store_result` stores the return value from `func_t f` to `*results` which is `return_t results[thread_work_size]`.

Comment (crcrpar) [reply aten/src/ATen/native/cuda/MemoryAccess.cuh:241]: In short, `store`'s functionality is not changed at all in this PR.

File: aten/src/ATen/native/cuda/CUDALoops.cuh [hunk 7]
```diff
@@ -42,6 +42,8 @@
 #include <c10/util/TypeCast.h>
 #include <c10/util/C++17.h>
 
+#include <thrust/tuple.h>
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/CUDALoops.cuh:45]: All your changes in this file should go to Loops.cuh instead of CUDALoops.cuh, because we will need ROCm support as well, and this file is not included on ROCm.

File: aten/src/ATen/native/cuda/CUDALoops.cuh [hunk 8]
```diff
@@ -296,4 +315,29 @@ void gpu_kernel_impl(TensorIterator& iter, const func_t& f) {
...
+  TORCH_INTERNAL_ASSERT(iter.ntensors() == ntensors);
+
+  at::detail::Array<char*, ntensors> data;
+  for (int i = 0; i < ntensors; i++) {
+    data[i] = (char*)iter.data_ptr(i);
+  }
+
+  int64_t numel = iter.numel();
+
+  auto input_calc = make_input_offset_calculator<num_inputs>(iter);
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/CUDALoops.cuh:338]: I believe there should be a logic
```C++
if (contiguous) {
   TrivialOffsetCalculator ...
else {
   ....
}
```

File: aten/src/ATen/native/cuda/MemoryAccess.cuh [hunk 9]
```diff
@@ -193,6 +214,28 @@ struct vectorized {
   }
 };
 
+template <typename data_t, typename inp_calc_t, typename out_calc_t, int num_outputs>
+struct multi_outputs_unroll : unroll<data_t, inp_calc_t, out_calc_t, num_outputs> {
+
+  __device__ multi_outputs_unroll(data_t data, int remaining, inp_calc_t ic, out_calc_t oc):
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/MemoryAccess.cuh:220]: Does `using unroll<data_t, inp_calc_t, out_calc_t, num_outputs>::unroll;` or `using unroll::unroll;` work?

File: aten/src/ATen/native/cuda/MemoryAccess.cuh [hunk 10]
```diff
@@ -193,6 +214,28 @@ struct vectorized {
...
+    unroll<data_t, inp_calc_t, out_calc_t, num_outputs>({data, remaining, ic, oc}) {}
+
+  template <typename return_t>
+  __device__ inline void store(return_t *from, int idx) {
+    int thread_idx = threadIdx.x;
+    #pragma unroll
+    for (int i = 0; i < thread_work_size; i++) {
+      if (thread_idx >= this->remaining) {
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/MemoryAccess.cuh:228]: Why are you using a lot of `this->`? Is it a good coding style?

File: aten/src/ATen/native/cuda/CUDALoops.cuh (hunk 7 above)
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/CUDALoops.cuh:45]: Can you revert the change in this file? We don't need these two lines.

File: aten/src/ATen/native/cuda/Loops.cuh [hunk 11]
```diff
@@ -127,4 +133,76 @@ void gpu_kernel_with_scalars(TensorIterator& iter, const func_t& f) {
...
+  auto stream = at::cuda::getCurrentCUDAStream();
+  unrolled_elementwise_kernel_for_multi_outputs<num_outputs, func_t, array_t><<<grid, num_threads, 0, stream>>>(N, f, data, ic, oc);
+  AT_CUDA_CHECK(cudaGetLastError());
+}
+
+template <typename func_t>
+void gpu_kernel_multiple_outputs_impl(TensorIterator& iter, const func_t& f) {
+  using traits = function_traits<func_t>;
+  using output_t = typename traits::result_type;
+  TORCH_INTERNAL_ASSERT(memory::detail::is_tuple<output_t>::value);
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/Loops.cuh:158]: does `static_assert` work?

File: aten/src/ATen/native/cuda/Loops.cuh [hunk 12]
```diff
@@ -127,4 +133,76 @@ void gpu_kernel_with_scalars(TensorIterator& iter, const func_t& f) {
...
+  constexpr int num_outputs = thrust::tuple_size<output_t>::value;
+  TORCH_INTERNAL_ASSERT(num_outputs > 1);
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/Loops.cuh:160]: Do we really need this? If returning a tuple of size 1 is working, then I don't think we need to check this.

File: aten/src/ATen/native/cuda/Loops.cuh [hunk 13]
```diff
@@ -127,4 +133,76 @@ void gpu_kernel_with_scalars(TensorIterator& iter, const func_t& f) {
...
+  constexpr int num_inputs = traits::arity;
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/Loops.cuh:161]: Just a note: this will be changed to `traits::arity()`, depending on when https://github.com/pytorch/pytorch/pull/35774 is landed.

File: aten/src/ATen/native/cuda/MemoryAccess.cuh [hunk 14]
```diff
@@ -64,15 +66,34 @@ struct vectorized_load_helper {
...
+template <typename ...T> struct is_tuple<thrust::tuple<T...>>: std::true_type {};
+
+template <int current>
+struct multi_outputs_store_helper {
+  template<int ntensors, int num_outputs, typename out_t>
+  C10_HOST_DEVICE static void apply(
+      at::detail::Array<char*, ntensors> data,
+      at::detail::Array<uint32_t, num_outputs> offsets,
+      out_t ret) {
+    static_assert(is_tuple<out_t>::value);
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/MemoryAccess.cuh:90]: Would it be better to use
```C++
  template<int ntensors, int num_outputs, typename ...Args>
  C10_HOST_DEVICE static void apply(
      at::detail::Array<char*, ntensors> data,
      at::detail::Array<uint32_t, num_outputs> offsets,
      thrust::tuple<Args...> ret) {
    // no static assert here
```

File: aten/src/ATen/native/cuda/Loops.cuh [hunk 15]
```diff
@@ -41,21 +43,25 @@ template<int N>
...
   return OffsetCalculator<N>(iter.ndim(), iter.shape().data(), strides.data(), element_sizes);
 }
 
-static OffsetCalculator<1> make_output_offset_calculator(const TensorIterator& iter) {
-  std::array<const int64_t*, 1> strides;
-  strides[0] = iter.strides(0).data();
-  int64_t element_size = iter.element_size(0);
-  return OffsetCalculator<1>(iter.ndim(), iter.shape().data(), strides.data(), &element_size);
+template <int num_outputs = 1>
+static OffsetCalculator<num_outputs> make_output_offset_calculator(const TensorIterator& iter) {
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/Loops.cuh:57]:
```C++
TORCH_INTERNAL_ASSERT(num_outputs == iter.noutputs());
```

File: aten/src/ATen/native/cuda/Loops.cuh [hunk 16]
```diff
@@ -127,4 +133,75 @@ void gpu_kernel_with_scalars(TensorIterator& iter, const func_t& f) {
...
+  auto stream = at::cuda::getCurrentCUDAStream();
+  unrolled_elementwise_kernel_for_multi_outputs<num_outputs, func_t, array_t><<<grid, num_threads, 0, stream>>>(N, f, data, ic, oc);
+  AT_CUDA_CHECK(cudaGetLastError());
+}
+
+template <typename func_t>
+void gpu_kernel_multiple_outputs_impl(TensorIterator& iter, const func_t& f) {
+  using traits = function_traits<func_t>;
+  using output_t = typename traits::result_type;
+  static_assert(memory::detail::is_tuple<output_t>::value);
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/Loops.cuh:158]: On C++14, `static_assert` requires you to leave a message. Compilers in Linux does not enforce that, but windows does.

File: aten/src/ATen/native/cuda/MemoryAccess.cuh [hunk 17]
```diff
@@ -64,15 +66,33 @@ struct vectorized_load_helper {
...
     // `data` hold the data_ptr for tensors [output, input0, input1, ...], so we
     // need a +1 offset to get the input
-    auto ptr = reinterpret_cast<arg_t *>(self.data[arg_index + 1]) + offset[arg_index];
+    auto ptr = reinterpret_cast<arg_t *>(self.data[arg_index + num_outputs]) + offset[arg_index];
     std::get<arg_index>(args[j]) = *ptr;
   }
 };
 
+// check the return type is `thrust::tuple`, not `std::tuple`.
+template <typename T> struct is_tuple: std::false_type {};
```
Comment (zasdfgbnm) [inline aten/src/ATen/native/cuda/MemoryAccess.cuh:79]: `is_tuple` is no longer needed in this file, could you move it to Loops.cuh?

**Comment by crcrpar** [comment]
Thank you for your review. As I filed the PR to upstream, closing.
